### ISSUE TITLE
[Phase 5] Query-by-Committee

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/multi-label-doc-classifier#6]
## Summary

Implement query-by-committee active learning using an ensemble of diverse classifiers. Samples where committee members disagree most are selected for labeling.

**Parent Issue**: #1
**Depends On**: Phase 4

## Objectives

- Build committee of diverse classifiers
- Implement disagreement metrics (vote entropy, KL divergence)
- Create hybrid strategies combining uncertainty and committee
- Ensure committee diversity through different base estimators

## System Context

```
src/
└── active_learning/
    ├── committee.py       # NEW - Query-by-committee implementation
    ├── disagreement.py    # NEW - Disagreement metrics
    └── hybrid.py          # NEW - Combined strategies
```

### Query-by-Committee
- [ ] Create `src/active_learning/committee.py`
- [ ] Implement `QueryByCommittee` class
- [ ] Train multiple diverse classifiers
- [ ] Support different base estimator types
- [ ] Bootstrap sampling for training diversity

### Disagreement Metrics
- [ ] Create `src/active_learning/disagreement.py`
- [ ] Implement `vote_entropy`: Entropy of committee votes
- [ ] Implement `kl_divergence`: Average KL divergence between members
- [ ] Implement `disagreement_ratio`: Fraction of disagreeing members

### Hybrid Strategies
- [ ] Create `src/active_learning/hybrid.py`
- [ ] Combine uncertainty and committee scores
- [ ] Weighted combination with tunable parameters
- [ ] Alternating strategy (switch between methods)

### QueryByCommittee

```python
# src/active_learning/committee.py
import numpy as np
from sklearn.base import clone
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier
from typing import List, Optional

class QueryByCommittee:
    """Query-by-committee active learning with diverse classifiers."""

    def __init__(
        self,
        n_members: int = 5,
        base_estimators: Optional[List] = None,
        bootstrap: bool = True,
        random_state: Optional[int] = None
    ):
        """Initialize committee.

        Args:
            n_members: Number of committee members
            base_estimators: List of sklearn estimators (will cycle if fewer than n_members)
            bootstrap: Use bootstrap sampling for training diversity
            random_state: Random seed
        """
        self.n_members = n_members
        self.bootstrap = bootstrap
        self.random_state = random_state
        self.rng = np.random.RandomState(random_state)

        # Default diverse estimators
        if base_estimators is None:
            self.base_estimators = [
                LogisticRegression(max_iter=1000, C=1.0),
                LogisticRegression(max_iter=1000, C=0.1),
                LogisticRegression(max_iter=1000, C=10.0),
                DecisionTreeClassifier(max_depth=10),
                DecisionTreeClassifier(max_depth=5),
            ]
        else:
            self.base_estimators = base_estimators

        self.committee_ = []

    def fit(self, X, y):
        """Train all committee members.

        Args:
            X: Feature matrix
            y: Label matrix
        """
        self.committee_ = []
        n_samples = X.shape[0]

        for i in range(self.n_members):
            # Cycle through base estimators
            estimator = clone(self.base_estimators[i % len(self.base_estimators)])

            if self.bootstrap:
                # Bootstrap sample for diversity
                indices = self.rng.choice(n_samples, size=n_samples, replace=True)
                X_boot = X[indices]
                y_boot = y[indices]
            else:
                X_boot, y_boot = X, y

            estimator.fit(X_boot, y_boot)
            self.committee_.append(estimator)

        return self

    def predict_all(self, X) -> np.ndarray:
        """Get predictions from all committee members.

        Returns:
            Array of shape (n_members, n_samples, n_labels)
        """
        predictions = []
        for member in self.committee_:
            pred = member.predict(X)
            predictions.append(pred)
        return np.array(predictions)

    def predict_proba_all(self, X) -> np.ndarray:
        """Get probability predictions from all committee members.

        Returns:
            Array of shape (n_members, n_samples, n_labels)
        """
        probas = []
        for member in self.committee_:
            if hasattr(member, 'predict_proba'):
                proba = member.predict_proba(X)
                # Handle sklearn's format for binary classification
                if isinstance(proba, list):
                    # Multi-label case: list of arrays per label
                    proba = np.column_stack([p[:, 1] if p.ndim > 1 else p for p in proba])
            else:
                proba = member.predict(X).astype(float)
            probas.append(proba)
        return np.array(probas)

    def predict(self, X) -> np.ndarray:
        """Majority vote prediction across committee."""
        all_preds = self.predict_all(X)
        # Vote: if more than half predict 1, predict 1
        return (np.mean(all_preds, axis=0) >= 0.5).astype(int)

    def compute_disagreement(self, X, method: str = 'vote_entropy') -> np.ndarray:
        """Compute disagreement scores for each sample.

        Args:
            X: Feature matrix
            method: 'vote_entropy', 'kl_divergence', or 'disagreement_ratio'

        Returns:
            Disagreement scores (higher = more disagreement)
        """
        all_preds = self.predict_all(X)  # (n_members, n_samples, n_labels)

        if method == 'vote_entropy':
            return self._vote_entropy(all_preds)
        elif method == 'kl_divergence':
            all_probas = self.predict_proba_all(X)
            return self._kl_divergence(all_probas)
        elif method == 'disagreement_ratio':
            return self._disagreement_ratio(all_preds)
        else:
            raise ValueError(f"Unknown method: {method}")

    def _vote_entropy(self, all_preds: np.ndarray) -> np.ndarray:
        """Entropy of vote distribution across committee members."""
        # all_preds: (n_members, n_samples, n_labels)
        n_members = all_preds.shape[0]

        # Count votes for each label being 1
        vote_counts = np.sum(all_preds, axis=0)  # (n_samples, n_labels)
        vote_ratio = vote_counts / n_members

        # Compute entropy for each label
        vote_ratio = np.clip(vote_ratio, 1e-10, 1 - 1e-10)
        entropy = -vote_ratio * np.log2(vote_ratio) - (1 - vote_ratio) * np.log2(1 - vote_ratio)

        # Average across labels
        return np.mean(entropy, axis=1)

    def _disagreement_ratio(self, all_preds: np.ndarray) -> np.ndarray:
        """Fraction of labels where committee disagrees."""
        n_members = all_preds.shape[0]

        # For each sample and label, check if all members agree
        vote_counts = np.sum(all_preds, axis=0)  # (n_samples, n_labels)
        disagreement = (vote_counts > 0) & (vote_counts < n_members)

        # Fraction of labels with disagreement
        return np.mean(disagreement, axis=1)

    def _kl_divergence(self, all_probas: np.ndarray) -> np.ndarray:
        """Average pairwise KL divergence between committee members."""
        n_members, n_samples, n_labels = all_probas.shape

        # Compute mean prediction
        mean_proba = np.mean(all_probas, axis=0)  # (n_samples, n_labels)
        mean_proba = np.clip(mean_proba, 1e-10, 1 - 1e-10)

        # KL divergence from each member to mean
        total_kl = np.zeros(n_samples)
        for i in range(n_members):
            member_proba = np.clip(all_probas[i], 1e-10, 1 - 1e-10)
            # Binary KL divergence
            kl = member_proba * np.log(member_proba / mean_proba) + \
                 (1 - member_proba) * np.log((1 - member_proba) / (1 - mean_proba))
            total_kl += np.mean(kl, axis=1)

        return total_kl / n_members

    def select_samples(self, X, n_samples: int, method: str = 'vote_entropy') -> np.ndarray:
        """Select samples with highest disagreement."""
        scores = self.compute_disagreement(X, method)
        return np.argsort(scores)[-n_samples:][::-1]
```

### Hybrid Strategy

```python
# src/active_learning/hybrid.py
import numpy as np
from typing import Optional
from .uncertainty import UncertaintySampler
from .committee import QueryByCommittee

class HybridSampler:
    """Combines uncertainty sampling and query-by-committee."""

    def __init__(
        self,
        uncertainty_weight: float = 0.5,
        uncertainty_strategy: str = 'entropy',
        committee_method: str = 'vote_entropy',
        n_committee_members: int = 5
    ):
        self.uncertainty_weight = uncertainty_weight
        self.uncertainty_sampler = UncertaintySampler(strategy=uncertainty_strategy)
        self.committee_method = committee_method
        self.committee = QueryByCommittee(n_members=n_committee_members)

    def fit(self, X, y):
        """Fit the committee classifiers."""
        self.committee.fit(X, y)
        return self

    def compute_scores(self, classifier, X_pool: np.ndarray) -> np.ndarray:
        """Compute combined uncertainty + committee scores."""
        # Normalize both to [0, 1] range
        uncertainty_scores = self.uncertainty_sampler.compute_scores(classifier, X_pool)
        uncertainty_scores = (uncertainty_scores - uncertainty_scores.min()) / \
                            (uncertainty_scores.max() - uncertainty_scores.min() + 1e-10)

        committee_scores = self.committee.compute_disagreement(X_pool, self.committee_method)
        committee_scores = (committee_scores - committee_scores.min()) / \
                          (committee_scores.max() - committee_scores.min() + 1e-10)

        # Weighted combination
        combined = (self.uncertainty_weight * uncertainty_scores +
                   (1 - self.uncertainty_weight) * committee_scores)

        return combined

    def select_samples(self, classifier, X_pool: np.ndarray, n_samples: int) -> np.ndarray:
        """Select samples using combined scoring."""
        scores = self.compute_scores(classifier, X_pool)
        return np.argsort(scores)[-n_samples:][::-1]
```

## Verification

```python
from src.active_learning.committee import QueryByCommittee

# Train committee
committee = QueryByCommittee(n_members=5)
committee.fit(X_train, y_train)

# Verify diversity: members should make different predictions
all_preds = committee.predict_all(X_test)
for i in range(committee.n_members):
    for j in range(i + 1, committee.n_members):
        diff = np.mean(all_preds[i] != all_preds[j])
        print(f"Members {i} vs {j}: {diff:.2%} different")
        assert diff > 0.01, "Committee members should make different predictions"

# Verify disagreement scoring
disagreement = committee.compute_disagreement(X_test, method='vote_entropy')
assert np.all(disagreement >= 0), "Disagreement should be non-negative"
assert np.any(disagreement > 0), "Some samples should have disagreement"

# Active learning should outperform random
# ... comparison code ...
```

## Definition of Done

- [ ] Committee trains successfully with diverse estimators
- [ ] Committee members make different predictions (>5% difference)
- [ ] All three disagreement metrics compute correctly
- [ ] Hybrid strategy combines scores meaningfully
- [ ] Query-by-committee outperforms random sampling
- [ ] Unit tests achieve 80% coverage

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/active_learning/committee.py` | - | Create | QBC implementation |
| `src/active_learning/disagreement.py` | - | Create | Disagreement metrics |
| `src/active_learning/hybrid.py` | - | Create | Combined strategies |
| `tests/test_committee.py` | - | Create | Committee tests |
| `tests/test_hybrid.py` | - | Create | Hybrid strategy tests |
